### ISSUE TITLE
ensure Linux compatibility

### ISSUE THREAD

**Issue by moonshoes87** [PmagPy/PmagPy#346]
get Linux distribution set up, for standalones and general install instructions

**Comment by moonshoes87** [comment]
I've been trying to get a Linux VM set up, but I'm having no success installing wxPython.  I've tried following the [docs](https://github.com/wxWidgets/Phoenix/blob/master/README.rst#how-to-build-phoenix), but I'm getting nothing but horrible and lengthy error messages.

@Caoimhinmg, any insight into the wxPython install process?

**Comment by moonshoes87** [comment]
I've gone through a few error messages, but this is the tail end of the one I can't seem to get past:

    g++: internal compiler error: Segmentation fault (program cc1plus)
    Please submit a full bug report,
    with preprocessed source if appropriate.
    See <file:///usr/share/doc/gcc-5/README.Bugs> for instructions.
    Makefile:38646: recipe for target 'advdll_animatecmn.o' failed
    make: *** [advdll_animatecmn.o] Error 4
    make: *** Waiting for unfinished jobs....
    (cd /tmp/pip-install-br112omt/wxPython/build/wxbld/gtk3/lib/; rm -f libwx_baseu-3.0.so libwx_baseu-3.0.so.0; ln -s libwx_baseu-3.0.so.0.4.0 libwx_baseu-3.0.so.0; ln -s libwx_baseu-3.0.so.0 libwx_baseu-3.0.so)
    Error building
    ERROR: failed building wxWidgets
    Traceback (most recent call last):
      File "build.py", line 1321, in cmd_build_wx
        wxbuild.main(wxDir(), build_options)
      File "/tmp/pip-install-br112omt/wxPython/buildtools/build_wxwidgets.py", line 498, in main
        exitIfError(wxBuilder.build(dir=buildDir, options=args), "Error building")
      File "/tmp/pip-install-br112omt/wxPython/buildtools/build_wxwidgets.py", line 85, in exitIfError
        raise builder.BuildError(msg)
    buildtools.builder.BuildError: Error building
    Finished command: build_wx (0.623s)
    Finished command: build (0.623s)
    Command '"/home/parallels/anaconda3/bin/python" -u build.py build' failed with exit code 1.
    
    ----------------------------------------
    Command "/home/parallels/anaconda3/bin/python -u -c "import setuptools, tokenize;__file__='/tmp/pip-install-br112omt/wxPython/setup.py';f=getattr(tokenize, 'open', open)(__file__);code=f.read().replace('\r\n', '\n');f.close();exec(compile(code, __file__, 'exec'))" install --record /tmp/pip-record-z25q6k6c/install-record.txt --single-version-externally-managed --compile" failed with error code 1 in /tmp/pip-install-br112omt/wxPython/

**Comment by Caoimhinmg** [comment]
Hey @moonshoes87, first things first. What kind of Linux you using? Ubuntu, Mint, Red Hat, Arch (god forbid)? Need to know what we're working with.

Have you used a package manager like apt or yum to install all of the packages mentioned on that docs page under the Linux header near the bottom? Including the GTK2 ones just in case? Then tried the install command again? It looks like it's failing when building the C based WxWidgets library which probably means theres a missing dependency.

**Comment by moonshoes87** [comment]
Ok, I got past this error.  I am going to add the instructions for building up PmagPy on Ubuntu from scratch.

**Comment by moonshoes87** [comment]
@Caoimhinmg, sorry for the lack of details.  I was using Ubuntu 16.04 in a Parallels VM.  I've added the list of steps I used to the repo: https://github.com/PmagPy/PmagPy/blob/master/setup_scripts/linux_instructions.md

It may or may not be enough to actually get someone else set up, but it should at least point them in the right direction.

**Comment by moonshoes87** [comment]
I'm planning to start doing Linux standalone releases in addition to OS X and Windows.

**Comment by moonshoes87** [comment]
Everything is working on my Linux VM, except that some of the GUI tests crash.  The functionality that is tested by those tests does work, though.

Pyinstaller is also working, and I'm seeing if I can get UPX.

**Comment by moonshoes87** [comment]
sudo apt-get install upx

**Comment by Caoimhinmg** [comment]
@moonshoes87 the GUI tests always were finicky anyway. I just checked and I think the package you want is upx-ucl. Not super sure though I never used this utility. If you aren't sure what you're looking for in the Ubuntu repositories ever use this command: `sudo apt-cache search $KEYWORD`

**Comment by moonshoes87** [comment]
Thanks!  Upx seems to be working.  However, my executables aren’t working correctly.  Pyinstaller runs to completion, and the resulting executable works if run them from the command line but not if I run them by double-clicking the icon.  This seems rather unusual and I’m not sure how to troubleshoot it yet.  Have you encountered this?

**Comment by Caoimhinmg** [comment]
Have you tried changing permissions? `chmod a+x $FILENAME`
If that doesn't work check what it's running when you double click and make sure it's using something sensible like autorunprompt or mono.
Finally when you run in command line make sure that you are running the local binary using ./ not accidentally running something else in path so you can debug it.

**Comment by moonshoes87** [comment]
Permissions don’t make a difference, and I am definitely running the local version.

How do you check what it’s using to run?  Looking in System monitor, I just see the name of the executable, i.e. pmag_gui_3.19.0.

**Comment by moonshoes87** [comment]
I’m trying pyinstaller with a minimal wxpython script and seeing what I get.

**Comment by Caoimhinmg** [comment]
right click on it and click "open with" you should be able to see what it's using to run it and you should be able to pick which program to use.

**Comment by moonshoes87** [comment]
Hmmm, it doesn’t have any default set, so maybe that is the problem.  And I don’t have either of the programs you suggested.  It looks like I can install mono using apt-get, but that downloads a crap-ton of packages.

This is a similar issue, but it is being reported on OS X:  https://github.com/pyinstaller/pyinstaller/issues/1804

**Comment by moonshoes87** [comment]
This works:
 
    parallels@parallels-vm:~/Desktop$ ./dist/simple 

This fails:

    parallels@parallels-vm:~/Desktop$ open ./dist/simple 
    Couldn't get a file descriptor referring to the console

**Comment by moonshoes87** [comment]
When I open the last Linux executable that you released, it opens and runs just fine.  However, when I right-click it just says “Run” and doesn’t specify the program for that either, so I’m not sure how it works even when it’s working, hah.

If I rename my executable from simple to simple.exe, it opens the Archive Viewer instead.  Renaming to simple.app and nothing happens (as before).

Adding these lines to the .spec file doesn’t do anything:

    app = BUNDLE(exe,
             #name='pmag_gui.app',
             name='simple' + ".app",
             bundle_identifier=None)

**Comment by Caoimhinmg** [comment]
Okay that is honestly very strange and usually Linux is nothing if not consistently logical in it's stupidity. Tomorrow I'll see what happens when I try this myself and try to remember my control flow for making the binaries to see if I can identify where we are differing.

**Comment by moonshoes87** [comment]
Thanks -- I'm definitely confused at this point.  Let me know what you find.

**Comment by Caoimhinmg** [comment]
I just made a binary using pyinstaller 3.3.1 and it works fine. I just used the spec files from [here](https://github.com/PmagPy/PmagPy-Standalone-Linux) dropped into the main directory. I'm not sure what we could be doing different to cause this problem. If you send me your binary I could try running it over here to see if it's just your system?

**Comment by moonshoes87** [comment]
I initially used the same [spec file](https://github.com/PmagPy/PmagPy/blob/master/pmag_gui.spec) from the repo I use for both OS X and Windows.

I then tried with commenting out the last few lines:

#app = BUNDLE(exe,
#             #name='pmag_gui.app',
#             name=app_name + ".app",
#             icon='./programs/images/PmagPy.ico',
#             bundle_identifier=None)

But that doesn't work either.  It doesn't seem substantially different from your .spec file.

I can't attach or send the executable easily because it's too big. Github won't allow it in a comment, and email won't work. I could share it with you using Dropbox. Or I could make a (presumably) non-functional release to PmagPy-Standalone-Linux, which you could then replace with a working executable if necessary.  Let me know what you prefer.

**Comment by Caoimhinmg** [comment]
Let's roll with a google drive or drop box link so we don't clutter the version history. I just tried it using your spec file and it worked fine as well. I imagine that this is more an issue of your platform and how it chooses to handle the executable when you double click it than with the executable itself.

**Comment by moonshoes87** [comment]
Great, here's a link: https://drive.google.com/open?id=1h14zF7m8kgNHbK_7IrQmySTCJ4ZmrXxQ

**Comment by Caoimhinmg** [comment]
Ah, yes I had forgotten about these darn bugs. @moonshoes87 your problem is not pyinstaller or your system.

You're not bundling everything in the executable or it's not freezing something. When you double click it opens in a different environment and has different paths so it fails when you try it on your machine. When you run it in terminal you must be doing it somewhere where it can find the needed dependencies and so it works fine. As it stands the binary is running when you double click it, it just errors before it launches. You can see this by running top in terminal then double clicking it and watching as the process pmag_gui_3.19.0 shows up then goes away just as fast.

That said on my machine it cannot find any of the paths ever so I get an error in terminal:
```bash
-I- Importing Pmag GUI dependencies
Traceback (most recent call last):
  File "/home/parallels/anaconda3/lib/python3.6/site-packages/PyInstaller/loader/pyimod03_importers.py", line 712, in load_module
  File "<frozen importlib._bootstrap_external>", line 399, in _check_name_wrapper
  File "<frozen importlib._bootstrap_external>", line 823, in load_module
  File "<frozen importlib._bootstrap_external>", line 682, in load_module
  File "<frozen importlib._bootstrap>", line 265, in _load_module_shim
  File "<frozen importlib._bootstrap>", line 684, in _load
  File "<frozen importlib._bootstrap>", line 658, in _load_unlocked
  File "<frozen importlib._bootstrap>", line 571, in module_from_spec
  File "<frozen importlib._bootstrap_external>", line 922, in create_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
ImportError: libwx_gtk3u_core-3.0.so.0: cannot open shared object file: No such file or directory

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "programs/pmag_gui.py", line 11, in <module>
  File "/home/parallels/anaconda3/lib/python3.6/site-packages/PyInstaller/loader/pyimod03_importers.py", line 628, in exec_module
  File "site-packages/wx/__init__.py", line 17, in <module>
  File "/home/parallels/anaconda3/lib/python3.6/site-packages/PyInstaller/loader/pyimod03_importers.py", line 628, in exec_module
  File "site-packages/wx/core.py", line 12, in <module>
  File "/home/parallels/anaconda3/lib/python3.6/site-packages/PyInstaller/loader/pyimod03_importers.py", line 714, in load_module
ImportError: libwx_gtk3u_core-3.0.so.0: cannot open shared object file: No such file or directory: wx._core
[21064] Failed to execute script pmag_gui
```

Not really sure how to fix this off the top of my head, might look into libwx_gtk3u_core-3.0.so.0 try searching for parts of this in apt and you might find some library that's missing.

In short it's got to be an environment thing, probably caused by an unsatisfied/unfrozen dependency that's only working in your one case because you built it there, or that's my best guess. Let me know anything else I can test, or if you want me to do a quick Linux release while you figure out the last few bugs in your new VM.

**Comment by moonshoes87** [comment]
Great, this gives me somewhere to start.  Thanks.  I'll let you know if I need another round of debugging help.  
 
It seems like the program *should* be finding wx._core successfully.  In EXE-00.toc:

      ('wx._core',
       '/home/parallels/anaconda3/lib/python3.6/site-packages/wx/_core.cpython-36m-x86_64-linux-gnu.so',
       'EXTENSION'),

which is right.  I wonder if the problem is related to this issue where I couldn't import wx:  https://github.com/pyenv/pyenv/issues/691.  I "solved" it by adding the following line to my .bashrc file:

   export LD_LIBRARY_PATH=~/anaconda3/lib/python3.6/site-packages/wx/

I tried installing:

    sudo apt-get install libwxgtk3.0

but that broke everything even worse, and now pyinstaller croaks on the "analyzing pmag_gui.py" step.   I've now uninstalled those packages, and pyinstaller still doesn't run.  pmag_gui.py still starts up just fine.  This is perplexing.  

I'm going to keep troubleshooting.

**Comment by moonshoes87** [comment]
I'm not sure this is really a high-priority issue.  I'm about to release 3.20.0, so @Caoimhinmg, if you wanted to do a Linux standalone release, that would be great.  I think I might need to start over with a clean VM, because there are so many variables as to what went wrong with my current VM.

**Comment by moonshoes87** [comment]
@Caoimhinmg, I'm trying again.  Fresh VM is now stalled out at the same place I got to last time.

Would you mind running

    pyinstaller pmag_gui.spec --onedir --noupx

    pyi-archive_viewer dist/pmag_gui_3.19.0 -b > output

and sending me the output?  I'd like to compare to see if I'm missing something.

**Comment by Caoimhinmg** [comment]
Okay, I ran the code you sent. Hope this helps. I just ran the code you sent on both the 3.19.0 binary that was made and a 3.20.0 binary that I think I made when I ran the first line you sent. Both binaries work over here.

[19_output.txt](https://github.com/PmagPy/PmagPy/files/2041697/output.txt)
[20_output.txt](https://github.com/PmagPy/PmagPy/files/2041698/output.txt)

**Comment by moonshoes87** [comment]
Thanks @Caoimhinmg.  I'll circle back to this as soon as I have some of Lisa's new bugs fixed.

**Comment by moonshoes87** [comment]
Things are currently working on Linux, I do believe.
